### ISSUE TITLE
Sprintf format not working on file input

### ISSUE THREAD

**Issue by pabrahamsson** [elasticsearch/logstash#2642]
I'm trying to use [sprintf format](http://www.elasticsearch.org/guide/en/logstash/current/configuration.html#_sprintf_format) for the input file but not having much luck.
Using the hard coded value for the input file (/var/log/tomcat6/localhost_access_log.2015-02-18.txt) works as expected and so does localhost_access_log.*.txt.

```
/opt/logstash/bin/logstash -e 'input { file { path => "/var/log/tomcat6/localhost_access_log.%{+yyyy-MM-dd}.txt" }}' --debug
Compiled pipeline code:
@inputs = []
@filters = []
@outputs = []
@input_file_1 = plugin("input", "file", LogStash::Util.hash_merge_many({ "path" => ("/var/log/tomcat6/localhost_access_log.%{+yyyy-MM-dd}.txt".force_encoding("UTF-8")) }))

@inputs << @input_file_1
@output_stdout_2 = plugin("output", "stdout", LogStash::Util.hash_merge_many({ "codec" => ("rubydebug".force_encoding("UTF-8")) }))

@outputs << @output_stdout_2
  @filter_func = lambda do |event, &block|
    extra_events = []
    @logger.debug? && @logger.debug("filter received", :event => event.to_hash)
    extra_events.each(&block)
  end
  @output_func = lambda do |event, &block|
    @logger.debug? && @logger.debug("output received", :event => event.to_hash)
    @output_stdout_2.handle(event)

  end {:level=>:debug, :file=>"logstash/pipeline.rb", :line=>"26"}
Using milestone 2 input plugin 'file'. This plugin should be stable, but if you see strange behavior, please let us know! For more information on plugin milestones, see http://logstash.net/docs/1.4.2-modified/plugin-milestones {:level=>:warn, :file=>"logstash/config/mixin.rb", :line=>"209"}
config LogStash::Codecs::Plain/@charset = "UTF-8" {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Inputs::File/@path = ["/var/log/tomcat6/localhost_access_log.%{+yyyy-MM-dd}.txt"] {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Inputs::File/@debug = false {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Inputs::File/@codec = <LogStash::Codecs::Plain charset=>"UTF-8"> {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Inputs::File/@add_field = {} {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Inputs::File/@stat_interval = 1 {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Inputs::File/@discover_interval = 15 {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Inputs::File/@sincedb_write_interval = 15 {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Inputs::File/@start_position = "end" {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Outputs::Stdout/@codec = <LogStash::Codecs::RubyDebug > {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Outputs::Stdout/@type = "" {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Outputs::Stdout/@tags = [] {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Outputs::Stdout/@exclude_tags = [] {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
config LogStash::Outputs::Stdout/@workers = 1 {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"105"}
Registering file input {:path=>["/var/log/tomcat6/localhost_access_log.%{+yyyy-MM-dd}.txt"], :level=>:info, :file=>"logstash/inputs/file.rb", :line=>"74"}
No sincedb_path set, generating one based on the file path {:sincedb_path=>"/root/.sincedb_348a84635b0bb2bb6333204a8a99368e", :path=>["/var/log/tomcat6/localhost_access_log.%{+yyyy-MM-dd}.txt"], :level=>:info, :file=>"logstash/inputs/file.rb", :line=>"115"}
Pipeline started {:level=>:info, :file=>"logstash/pipeline.rb", :line=>"78"}
_sincedb_open: reading from /root/.sincedb_348a84635b0bb2bb6333204a8a99368e {:level=>:debug, :file=>"filewatch/tail.rb", :line=>"199"}
_discover_file_glob: /var/log/tomcat6/localhost_access_log.%{+yyyy-MM-dd}.txt: glob is: [] {:level=>:debug, :file=>"filewatch/watch.rb", :line=>"117"}
_discover_file_glob: /var/log/tomcat6/localhost_access_log.%{+yyyy-MM-dd}.txt: glob is: [] {:level=>:debug, :file=>"filewatch/watch.rb", :line=>"117"}
_discover_file_glob: /var/log/tomcat6/localhost_access_log.%{+yyyy-MM-dd}.txt: glob is: [] {:level=>:debug, :file=>"filewatch/watch.rb", :line=>"117"}
^CInterrupt received. Shutting down the pipeline. {:level=>:warn, :file=>"logstash/agent.rb", :line=>"119"}
Sending shutdown signal to input thread {:thread=>#<Thread:0x2b5cf9de sleep>, :level=>:info, :file=>"logstash/pipeline.rb", :line=>"236"}
caller requested sincedb write () {:level=>:debug, :file=>"filewatch/tail.rb", :line=>"185"}
caller requested sincedb write () {:level=>:debug, :file=>"filewatch/tail.rb", :line=>"185"}
Plugin is finished {:plugin=><LogStash::Outputs::Stdout >, :level=>:info, :file=>"logstash/plugin.rb", :line=>"59"}
Pipeline shutdown complete. {:level=>:info, :file=>"logstash/pipeline.rb", :line=>"89"}
```

**Comment by untergeek** [comment]
I am sorry for the confusion.  Sprintf format will not work inside the input block.  The reason is because the values that are used to build `%{+yyyy-MM-dd}` are pulled from the `@timestamp` field.  This field does not exist inside the input block because events do not exist yet.  Events are only created at the transition from the input block to the filter block.  It is also for this reason (no fields yet exist to use for comparison) that conditionals also cannot be used within the input block.

As such, conditionals and sprintf formatting can only be used inside the filter and output blocks.

We should update our documentation to make this more clear.  I will open an issue for that.

**Comment by untergeek** [comment]
Addressed in #2660
